libs/backend/src/scan/cast_vote_records/cast_vote_record_exporter.ts

```diff
@@ -0,0 +1,415 @@
+  BallotIdSchema,
+  BallotPageLayout,
+  BatchInfo,
+  CVR,
+  ElectionDefinition,
+  PollsState,
+  unsafeParse,
+} from '@votingworks/types';
+import { UsbDrive } from '@votingworks/usb-drive';
+import {
+  BooleanEnvironmentVariableName,
+  generateCastVoteRecordExportDirectoryName,
+  generateElectionBasedSubfolderName,
+  isFeatureFlagEnabled,
+  SCANNER_RESULTS_FOLDER,
+} from '@votingworks/utils';
+
+import { ExportDataError, Exporter } from '../../exporter';
+import { SCAN_ALLOWED_EXPORT_PATTERNS, VX_MACHINE_ID } from '../globals';
+import { buildCastVoteRecord } from './build_cast_vote_record';
+import { buildCastVoteRecordReportMetadata } from './build_report_metadata';
+import {
+  CanonicalizedSheet,
+  canonicalizeSheet,
+  describeSheetValidationError,
+} from './canonicalize';
+import { ResultSheet } from './export';
+import { buildElectionOptionPositionMap } from './option_map';
+
+type ExportCastVoteRecordToUsbDriveError =
+  | { type: 'invalid-sheet-found'; message: string }
+  | ExportDataError;
+
+/**
+ * The subset of scanner store methods relevant to exporting cast vote records
+ */
+export interface ScannerStore {
+  clearCastVoteRecordHashes(): void;
+  getBatches(): BatchInfo[];
+  getCastVoteRecordRootHash(): string;
+  getDefiniteMarkThreshold(): number;
+  getElectionDefinition(): ElectionDefinition | undefined;
+  getExportDirectoryName(): string | undefined;
+  getMode(): 'live' | 'test';
+  getPollsState(): PollsState | undefined;
+  updateCastVoteRecordHashes(
+    castVoteRecordId: string,
+    castVoteRecordHash: string
+  ): void;
+  setExportDirectoryName(exportDirectoryName: string): void;
+}
+
+/**
+ * State that can be retrieved via the ScannerStore interface and that is unchanged by exporting
+ * cast vote records
+ */
+interface ScannerStateUnchangedByExport {
+  batches: BatchInfo[];
+  definiteMarkThreshold: number;
+  electionDefinition: ElectionDefinition;
+  inTestMode: boolean;
+  pollsState?: PollsState;
+}
+
+interface File {
+  fileName: string;
+  fileContents: string | Buffer;
+}
+
+/**
+ * Metadata stored in the top-level metadata file for a cast vote record export
+ */
+interface CastVoteRecordExportMetadata {
+  arePollsClosed?: boolean;
+  /** Global data relevant to all cast vote records in an export, e.g. election info */
+  castVoteRecordReportMetadata: CVR.CastVoteRecordReport;
+  /** A hash of all cast vote record files in an export */
+  castVoteRecordRootHash: string;
+}
+
+/**
+ * The API for a cast vote record exporter that a scanner can use to export cast vote records.
+ * Supports both continuous and batch export use cases.
+ */
+export interface CastVoteRecordExporterApi {
+  exportCastVoteRecordsToUsbDrive(
+    resultSheets: ResultSheet[] | Generator<ResultSheet>
+  ): Promise<Result<void, ExportCastVoteRecordToUsbDriveError>>;
+}
+
+/**
+ * The implementation of the cast vote record exporter API
+ */
+export class CastVoteRecordExporter implements CastVoteRecordExporterApi {
```

Comment:
It looks like this class doesn't actually hold any state in memory. Does it need to be its own object, or could it just be a function `exportCastVoteRecordsToUsbDrive` that takes some of these stateful objects as arguments?

The main motivation I could think of for making it an object would be to inject it at the top level of the app and then be able to mock it, but I've generally been pushing for as little intermediate mocking as possible in our backends (to try to make the tests more realistic).

Curious to hear your thoughts!

Reply:
Really good question!

I did notice that the class doesn't hold any state in memory, so you're right that it doesn't need to be a class and could just be a function with several inputs.

But I've been finding functions with so many inputs to be a little clunky, given how much you have to pass to them when you want to call them. With a class, you can at least pre-wire certain required inputs when you initialize the class, say in an `index.ts` file, and not have to worry about those from then on.

This benefit then extends to internal helpers, too, in that you don't have to continue passing around the same inputs over and over. Though I know this can be mitigated by grouping inputs in an interface.

That said, I definitely agree with the point about minimizing intermediate mocking in our backend tests and testing more completely. This made me realize that I was unnecessarily mocking the cast vote record exporter in VxScan backend tests when I could just initialize the real deal. Updated accordingly in https://github.com/votingworks/vxsuite/pull/3836/commits/d09c99d2a9a69bccad7047c08729bdde30aac0b9

Overall though, I'm definitely open to switching to plain functions, with well encapsulated inputs

Reply:
Yeah, I hear you on the annoyance of passing lots of inputs around through many functions.

In this case, would you need to pass anything except the `UsbDrive`? I think the `Store` is already available and the `ArtifactAuthenticator` could be constructed on demand.

One thing I notice about this pattern is that it encourages creating stateless singletons and then feeling like they need to be passed around instead of building them on demand. So the pattern then proliferates. When really these objects are just a bundle of code with preconfigured arguments.

I guess I'm not opposed to that approach, but it does make it a bit hard to know what's stateful and what's not. But this is mostly a style/readability concern, and my opinions are weakly held.

Reply:
> One thing I notice about this pattern is that it encourages creating stateless singletons and then feeling like they need to be passed around instead of building them on demand

Really good point!

I know I'm guilty of this with the `ArtifactAuthenticator`, though there was another reason at play there. I initially planned for it to be constructed as needed, but it needs certain env vars to configure itself, which aren't available in test environments. So I switched to a dependency injection approach to make it easier to mock in tests. In retrospect, I could've just set up the right env vars in tests. That's what I ended up doing for integration tests.

> the `ArtifactAuthenticator` could be constructed on demand

See the above haha. That said, in my next PR, I'll be revisiting the `ArtifactAuthenticator`, and I'm thinking of addressing the above while I'm at it, so that it can in fact be constructed on demand, or just turned into plain functions.

I'm convinced by this thread and am thinking of switching to plain functions in this PR, too. The "too many arguments" problem can be solved by grouping arguments meaningfully. I think the `ScannerStore` abstraction largely accomplishes that.

Reply:
Oh interesting puzzle with the env vars. I hadn't thought of that. Maybe having parameters that default to env vars could be a good approach? That way they can be overridden in test, but the consuming app code doesn't need to worry about managing the env vars.

Reply:
^ So that actually is how the `ArtifactAuthenticator` works ([relevant link](https://github.com/votingworks/vxsuite/blob/dffb657cad7f104fc8f2f345186eeea236dd33b2/libs/auth/src/artifact_authenticator.ts#L68-L72)) and how I test it ([relevant link](https://github.com/votingworks/vxsuite/blob/dffb657cad7f104fc8f2f345186eeea236dd33b2/libs/auth/src/artifact_authenticator.test.ts#L99-L101)).

The challenge in app-level tests, assuming that we were to construct the `ArtifactAuthenticator` wherever it was needed, is that it would exist embedded within other code, where we couldn't easily modify how it's constructed. Ways around this of course, which I'll explore!